### ISSUE TITLE
runtime: malloc instancing generic function seems to deadlock(ish) during benchmark

### ISSUE THREAD

**Issue by joeycumines** [golang/go#58744]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.20 linux/amd64
</pre>

~I believe I've also replicated it on windows, but I haven't verified the stack trace looks the same.~ On windows, I have experienced an issue with similar characteristics, but haven't managed to retrieve a stacktrace. I have managed to reliably replicate the issue on an M2 Macbook, using v1.20.1.

### Does this issue reproduce with the latest release?

Yes, I've replicated it on v1.20.1. The stack trace looks the same.

### What operating system and processor architecture are you using (`go env`)?

Note: I only just updated the go version.

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE="on"
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/joey/.cache/go-build"
GOENV="/home/joey/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/joey/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/joey/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.20.1"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/joey/dev/go-utilpkg/go.mod"
GOWORK="/home/joey/dev/go-utilpkg/go.work"
CGO_CFLAGS="-O2 -g"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-O2 -g"
CGO_FFLAGS="-O2 -g"
CGO_LDFLAGS="-O2 -g"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build3747613050=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?

I don't have a nice replicator, but it's an issue I ran into within a for-fun personal project, so I can share the code:

```sh
git clone https://github.com/joeycumines/go-utilpkg.git
cd go-utilpkg
git checkout dc3656b25a8e74fb38c60aed2d81f77b3e156d96
cd logiface/zerolog
while go test -timeout=15m -count=6 -benchmem -run=^$ -bench=^BenchmarkContextFieldType_Int$ 2>&1 >bench.txt; do date; sleep 1; done
less bench.txt
```

The stack trace seems to indicate it (could, possibly) be related to instancing the `newBuilder` generic function: https://github.com/joeycumines/go-utilpkg/blob/dc3656b25a8e74fb38c60aed2d81f77b3e156d96/logiface/logger.go#L349

This function is used as the `sync.Pool.New` field, and is implemented like:

```go
func newBuilder[E Event]() any {
	return new(Builder[E])
}
```

### What did you expect to see?

The benchmark run successfully.

### What did you see instead?

```
goos: linux
goarch: amd64
pkg: github.com/joeycumines/go-utilpkg/logiface/zerolog
cpu: Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
BenchmarkContextFieldType_Int/variant=baseline-16         	 7817979	       158.2 ns/op	     512 B/op	       1 allocs/op
BenchmarkContextFieldType_Int/variant=baseline-16         	 7773796	       158.2 ns/op	     512 B/op	       1 allocs/op
BenchmarkContextFieldType_Int/variant=baseline-16         	 8651400	       141.2 ns/op	     512 B/op	       1 allocs/op
BenchmarkContextFieldType_Int/variant=baseline-16         	 8568614	       140.7 ns/op	     512 B/op	       1 allocs/op
BenchmarkContextFieldType_Int/variant=baseline-16         	 7566774	       140.2 ns/op	     512 B/op	       1 allocs/op
BenchmarkContextFieldType_Int/variant=baseline-16         	 8818876	       136.5 ns/op	     512 B/op	       1 allocs/op
BenchmarkContextFieldType_Int/variant=generic-16          	14591700	        84.28 ns/op	     152 B/op	       5 allocs/op
BenchmarkContextFieldType_Int/variant=generic-16          	SIGQUIT: quit
PC=0x45544a m=12 sigcode=0

goroutine 0 [idle]:
runtime.pcvalueCacheKey(...)
	/usr/local/go/src/runtime/symtab.go:906
runtime.pcvalue({0x6c6e98?, 0x6d3020?}, 0x4d699a00?, 0x45522a?, 0x7fe44d699b70?, 0x1?)
	/usr/local/go/src/runtime/symtab.go:923 +0x4a fp=0x7fe44d6999d0 sp=0x7fe44d699900 pc=0x45544a
runtime.pcdatavalue({0x6c6e98?, 0x6d3020?}, 0x0?, 0x18bfb6b4ed389?, 0x7fe44d699a58?)
	/usr/local/go/src/runtime/symtab.go:1109 +0x4f fp=0x7fe44d699a10 sp=0x7fe44d6999d0 pc=0x4560cf
runtime.gentraceback(0x10000c000047900?, 0x7fe44d699dc0?, 0x427805?, 0xc00033f6c0, 0x4, 0xc000209c58, 0x20, 0x0, 0x702420?, 0x0)
	/usr/local/go/src/runtime/traceback.go:362 +0x19e5 fp=0x7fe44d699d68 sp=0x7fe44d699a10 pc=0x45e165
runtime.callers.func1()
	/usr/local/go/src/runtime/traceback.go:844 +0x4f fp=0x7fe44d699dd0 sp=0x7fe44d699d68 pc=0x45f88f
runtime.systemstack()
	/usr/local/go/src/runtime/asm_amd64.s:496 +0x49 fp=0x7fe44d699dd8 sp=0x7fe44d699dd0 pc=0x4689c9

goroutine 667 [running]:
runtime.systemstack_switch()
	/usr/local/go/src/runtime/asm_amd64.s:463 fp=0xc000209bb0 sp=0xc000209ba8 pc=0x468960
runtime.callers(0x7fe4640bf828?, {0xc000209c58?, 0x7fe44ee9cd28?, 0x3c?})
	/usr/local/go/src/runtime/traceback.go:843 +0x98 fp=0xc000209c18 sp=0xc000209bb0 pc=0x45f7d8
runtime.mProf_Malloc(0xc000072010, 0x10)
	/usr/local/go/src/runtime/mprof.go:420 +0x6a fp=0xc000209d90 sp=0xc000209c18 pc=0x42e28a
runtime.profilealloc(0xc000072010?, 0x10?, 0x10?)
	/usr/local/go/src/runtime/malloc.go:1290 +0x85 fp=0xc000209dc8 sp=0xc000209d90 pc=0x40e5e5
runtime.mallocgc(0x10, 0x594180, 0x1)
	/usr/local/go/src/runtime/malloc.go:1138 +0x685 fp=0xc000209e30 sp=0xc000209dc8 pc=0x40e105
runtime.newobject(0xc000132000?)
	/usr/local/go/src/runtime/malloc.go:1254 +0x27 fp=0xc000209e58 sp=0xc000209e30 pc=0x40e467
github.com/joeycumines/go-utilpkg/logiface.newBuilder[...]()
	/home/joey/dev/go-utilpkg/logiface/logger.go:420 +0x25 fp=0xc000209e78 sp=0xc000209e58 pc=0x561a85
github.com/joeycumines/go-utilpkg/logiface.(*loggerShared[...]).init(...)
	/home/joey/dev/go-utilpkg/logiface/logger.go:349

.. TRUNCATED, 63MB file, it's just those two lines ..

github.com/joeycumines/go-utilpkg/logiface.(*loggerShared[...]).init(...)
	/home/joey/dev/go-utilpkg/logiface/logger.go:349
	/home/joey/dev/go-utilpkg/logiface/logger.go:*** Test killed: ran too long (6m0s).
signal: killed
FAIL	github.com/joeycumines/go-utilpkg/logiface/zerolog	365.059s
```

**Comment by qiulaidongfeng** [comment]
GOOS/GOARCH=windows/amd64
go version : go version go1.20.1 windows/amd64
I didn't have a reproduction problem on my computer

**Comment by joeycumines** [comment]
Interesting, thanks for testing @qiulaidongfeng.

I've managed to replicate _some sort_ of issue on an m1 macbook, still testing.
I can't seem to replicate on windows, though, which is not what I expected. Still testing that too though, using:

```
@echo off
setlocal EnableDelayedExpansion

:loop
go test -timeout=5m -count=6 -benchmem -run=^$ -bench=^BenchmarkContextFieldType_Int$ 2>&1 >bench.txt && (
    echo %date% %time%
    timeout /t 1 /nobreak >nul
    goto :loop
)
```

**Comment by qiulaidongfeng** [comment]
I use virtualbox to reproduce the problem in GOOS=linux GOARCH=amd64
[bench.txt](https://github.com/golang/go/files/10838701/bench.txt)
I tested GOOS=windows GOARCH=amd64 for dozens of minutes, and finally saw

*** Test killed: ran too long (16m0s).

exit status 1

**Comment by joeycumines** [comment]
I've updated the windows reproducer script, it seems I put `2>&1` in the wrong place (I didn't get the stack trace either).

I'm running it now, perhaps it'll output a useful stack trace.

**Comment by dmitshur** [comment]
CC @golang/runtime.

**Comment by prattmic** [comment]
This sounds familiar to #58513.

cc @mdempsky

**Comment by mdempsky** [comment]
Yeah, I suspect this is a duplicate of #58513 too. traceback.go:362 from the reported stack trace is inside the FUNCDATA_InlTree loop, which is consistent with that issue.

The fix for that issue has just been submitted to both master and release-branch.go1.20 (to be included in 1.20.2). It would be great if you could build either of these from source and confirm whether the issue still repros. Thanks.

**Comment by joeycumines** [comment]
I've only tested on linux/amd64 with `release-branch.go1.20`, but nothing has blown up yet.

Might run it later tonight as well, but it looks promising. Thanks for the assistance, I guess close this if you'd like. I'll re-open if I reproduce it using a patched build.
